src/Illuminate/Foundation/Testing/Concerns/InteractsWithDatabase.php

```diff
@@ -45,27 +46,23 @@ protected function assertDatabaseHas($table, array $data, $connection = null)
      */
     protected function parseData($table, array $data)
     {
-        if (class_exists($table) && new $table() instanceof Model) {
-            /* @var $model Model */
-            $table::unguard();
-
-            $model = (new $table($data));
-
-            $casted = false;
+        if (!class_exists($table) and ! is_a($table,Model::class,true))
+        {
+            return $data;
+        }
 
-            foreach ($data as $key => $value) {
-                if ($model->hasCast($key) && $model->isAttributeValueCastedProperly($key, $value)) {
-                    $casted = true;
-                    break;
-                }
+        $model = (new $table);
+        foreach ( $data as $key => $value ) {
+            if ( $model->hasCast( $key ) ) {
+                $data[ $key ] = match ($model->getCasts()[$key]){
+                        'array' => is_string( $value ) ? json_decode( $value ) : $value,
+                        default => $value
+                };
```

Comment:
The goal of this PR is to handle the casting of `array` type, so I handled that using a `match` expression in case of needing to handle more cast types later.